### PR TITLE
Support HuggingFace Models Converted From tf2onnx in Python Script 

### PR DESCRIPTION
**Pull request by liuziyue**
**Description**: Describe your changes.
- Change TF converter from keras2onnx to tf2onnx. 
- Add new models to the list.
- Support pre-trained TF models from latest transformers. 
- Adjust attention and skiplayernormalization node fusion logic to speed up "bert", "roberta", "gpt2", "distilbert", "distilgpt2", "albert", "xlm" models from tf2onnx. 
- Limitation: 
        does not support large model. 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Support TF huggingface models converted from tf2onnx.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/huggingface_models.py [hunk 1]
```diff
@@ -62,12 +72,13 @@
     "albert-large-v2": (["input_ids"], 12, False, "bert"),
     "albert-xlarge-v2": (["input_ids"], 12, True, "bert"),
     #"albert-xxlarge-v2": (["input_ids"], 12, True, "bert"),
+    "albert-xxlarge-v2": (["input_ids"], 12, True, "bert"),
```
Comment (wangyems) [inline onnxruntime/python/tools/transformers/huggingface_models.py:75]: comment this line?

File: onnxruntime/python/tools/transformers/onnx_exporter.py [hunk 2]
```diff
@@ -234,7 +245,8 @@ def load_pretrained_model(model_name, config, cache_dir, custom_model_class, is_
 
     if model_class_name == "GPT2ModelNoPastState":
         if is_tf_model:
-            raise NotImplementedError("TFGPT2ModelNoPastState is currently not supported.")
+            # raise NotImplementedError("TFGPT2ModelNoPastState is currently not supported.")
```
Comment (wangyems) [inline onnxruntime/python/tools/transformers/onnx_exporter.py:248]: nit:remove?

File: onnxruntime/python/tools/transformers/onnx_exporter.py [hunk 3]
```diff
@@ -375,46 +401,88 @@ def export_onnx_model_from_tf(model_name, opset_version, use_external_data_forma
...
+        if not use_external_data_format:
+            Path(tf_internal_model_path).parent.mkdir(parents=True, exist_ok=True)
+
+        import tf2onnx, zipfile
+        tf2onnx.logging.set_level(tf2onnx.logging.ERROR)
+        specs = []
+        for name, value in example_inputs.items():
+            dims = [None] * len(value.shape)
+            specs.append(tf.TensorSpec(tuple(dims), value.dtype, name=name))
+        _, _ = tf2onnx.convert.from_keras(model,
```
Comment (wangyems) [inline onnxruntime/python/tools/transformers/onnx_exporter.py:465]: does tf2onnx support dynamic axis right now?

File: onnxruntime/python/tools/transformers/test/test_optimizer.py [hunk 4]
```diff
@@ -326,10 +326,16 @@ def test_huggingface_dialogpt_fusion(self):
...
+    def test_huggingface_bert_base_cased_from_tf2onnx(self):
+        self._test_optimizer_on_tf_model("bert-base-cased", [0, 12, 0, 0, 0, 0, 25], 1)
+        self._test_optimizer_on_tf_model("bert-base-cased", [0, 12, 0, 0, 0, 0, 25], 2)
+        self._test_optimizer_on_tf_model("bert-base-cased", [0, 12, 0, 0, 0, 0, 25], 3)
+
+    def test_huggingface_distilgpt2_from_tf2onnx(self):
+        self._test_optimizer_on_tf_model("distilgpt2", [0, 0, 0, 0, 0, 12, 1], 1)
+
+    def test_huggingface_albert_from_tf2onnx(self):
+        self._test_optimizer_on_tf_model("albert-base-v1", [0, 0, 0, 0, 0, 0, 25], 1)
```
Comment (wangyems) [inline onnxruntime/python/tools/transformers/test/test_optimizer.py:338]: maybe to update the test? according to the description..

Comment (liuziyue) [reply onnxruntime/python/tools/transformers/test/test_optimizer.py:338]: The reason I have not added them is that the results of those models are slightly bigger than our threshold compared to the original model. Is it ok if I increase the "rtol" in "validate_onnx_model()" to 1e-3?

File: onnxruntime/python/tools/transformers/onnx_exporter.py (hunk 3 above)
Comment (liuziyue) [reply onnxruntime/python/tools/transformers/onnx_exporter.py:465]: Yes it does!

File: onnxruntime/python/tools/transformers/test/test_optimizer.py (hunk 4 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/test/test_optimizer.py:338]: Maybe it's better to temporarily disable the validation check like https://github.com/microsoft/onnxruntime/blob/06928e6160838fb3521b3cfae4a9f6ba9608b7df/onnxruntime/python/tools/transformers/test/test_optimizer.py#L305 so that we keep track of which model have bigger diffs
If the error of optimized model increases by one order of magnitude compared with the original one, I think we might need to take a closer look

File: onnxruntime/python/tools/transformers/fusion_attention.py [hunk 5]
```diff
@@ -166,31 +166,22 @@ def create_attention_node(self, mask_index: str, q_matmul: NodeProto, k_matmul:
...
-        if out_size != hidden_size:
-            logger.debug(
-                f"Shape for weights of Q is {in_size, out_size}, which does not match hidden_size={hidden_size}")
-            return None
-
-        qkv_weight = np.stack((qw, kw, vw), axis=-2)
+        # All the matrices have the same shape. For 2d weights, the shapes would be [in_size, out_size]. 
+        # For 3d weights, shape would be [in_size, a, b] where a*b = out_size
+        in_size = qw.shape[0]
+        out_size = np.prod(qw.shape[1:])
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_attention.py:172]: It is better to add back the check:
```
 if out_size != hidden_size:
        return None
```

File: onnxruntime/python/tools/transformers/onnx_model_bert_tf.py [hunk 6]
```diff
@@ -360,9 +374,22 @@ def fuse_attention(self):
...
+        start_nodes.extend(layer_norm_nodes)
+
+        for normalize_node in start_nodes:
             # SkipLayerNormalization has two inputs, and one of them is the root input for attention.
+            if normalize_node.op_type == 'LayerNormalization':
+                add_before_layernorm = self.match_parent(normalize_node, 'Add', 0)
+                if add_before_layernorm is not None:
+                    normalize_node = add_before_layernorm
+                else:
+                    return
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model_bert_tf.py:392]: return => continue?

File: onnxruntime/python/tools/transformers/huggingface_models.py [hunk 7]
```diff
@@ -13,13 +13,24 @@
...
     # BERT
-    "bert-base-uncased": (["input_ids", "attention_mask", "token_type_ids"], 11, False, "bert"),
-    "bert-large-uncased": (["input_ids", "attention_mask", "token_type_ids"], 11, False, "bert"),
-    "bert-base-cased": (["input_ids", "attention_mask", "token_type_ids"], 11, False, "bert"),
+    "bert-base-uncased": (["input_ids", "attention_mask", "token_type_ids"], 12, False, "bert"),
+    "bert-large-uncased": (["input_ids", "attention_mask", "token_type_ids"], 12, False, "bert"),
+    "bert-base-cased": (["input_ids", "attention_mask", "token_type_ids"], 12, False, "bert"),
+    "bert-large-cased": (["input_ids", "attention_mask", "token_type_ids"], 12, False, "bert"),
+    "bert-base-multilingual-uncased": (["input_ids", "attention_mask", "token_type_ids"], 12, False, "bert"),
+    "bert-base-multilingual-cased": (["input_ids", "attention_mask", "token_type_ids"], 12, False, "bert"),
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/huggingface_models.py:21]: Previously, we removed a few bert models from the list since they are near duplicated. Can we comment those models instead of adding them back?

Otherwise, list of models will be very long when you type `python benchmark.py --help`.

File: onnxruntime/python/tools/transformers/fusion_attention.py (hunk 5 above)
Comment (liuziyue) [reply onnxruntime/python/tools/transformers/fusion_attention.py:172]: added back to line 184

**Comment by tianleiwu** [comment]
Can we add a test case in test_attention_fusion.py for attention from tf2onnx?
Also, please make sure this change will not break the fusion of BERT model from keras2onnx.

**Comment by liuziyue** [comment]
@tianleiwu 
> Can we add a test case in test_attention_fusion.py for attention from tf2onnx?
> Also, please make sure this change will not break the fusion of BERT model from keras2onnx.

Added test case for attention. 
For the BERT models from keras2onnx, it's broken by #6920 . I did a quick fix on the function signature and after that it works fine.

**Review by tianleiwu — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]
